File: hongguo/detail.go
package hongguo

import (
	"context"
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"sort"
	"strconv"
	"strings"
)

func ParseAppDetail(result map[string]any, seriesID string) (Detail, error) {
	detail := nestedMap(result, "data", "video_data")
	returnedID := mapString(detail, "series_id_str", "series_id")
	if returnedID != seriesID {
		return Detail{}, errors.New("红果 App 未返回所请求的剧集")
	}
	entry := Detail{Drama: DramaFromAny(detail, "短剧")}
	seen := map[string]bool{}
	episodes := map[int]bool{}
	for _, row := range anyList(detail["video_list"]) {
		video, _ := row.(map[string]any)
		videoID := mapString(video, "vid")
		index, err := strconv.Atoi(mapString(video, "vid_index"))
		if err != nil || index < 1 || !NumericID.MatchString(videoID) {
			return entry, errors.New("红果 App 分集编号或视频 ID 无效")
		}
		if identifier := mapString(video, "series_id"); identifier != "" && identifier != seriesID {
			return entry, errors.New("红果 App 返回了其他剧集的分集")
		}
		if seen[videoID] || episodes[index] {
			return entry, errors.New("红果 App 返回了重复分集")
		}
		seen[videoID], episodes[index] = true, true
		entry.Chapters = append(entry.Chapters, Chapter{
			ID: ChapterID(seriesID, videoID), Source: Source,
			Title: fmt.Sprintf("第%d集", index), VideoURL: MediaPlaceholder(videoID), CurrentEpisode: rawEpisode(index),
		})
	}
	sort.Slice(entry.Chapters, func(left, right int) bool {
		leftIndex, _ := strconv.Atoi(entry.Chapters[left].EpisodeString(left + 1))
		rightIndex, _ := strconv.Atoi(entry.Chapters[right].EpisodeString(right + 1))
		return leftIndex < rightIndex
	})
	total, _ := strconv.Atoi(mapString(detail, "episode_cnt"))
	if len(entry.Chapters) == 0 || total > len(entry.Chapters) {
		return entry, errors.New("红果 App 未返回完整分集，已尝试其他详情入口")
	}
	for index, chapter := range entry.Chapters {
		if chapter.EpisodeString(index+1) != strconv.Itoa(index+1) {
			return entry, errors.New("红果 App 分集列表不连续")
		}
	}
	return entry, nil
}

func ParseWebDetail(body, seriesID string) (Detail, error) {
	page := RouterLoaderMap(ParseRouterData(body), "detail_page", "detail_")
	detail, _ := page["seriesDetail"].(map[string]any)
	if len(detail) == 0 {
		return Detail{}, fmt.Errorf("hongguo detail is empty")
	}
	title := firstNonEmpty(mapString(detail, "series_name", "series_title", "name"), seriesID)
	vids := anyList(detail["vid_list"])
	chapters := make([]Chapter, 0, len(vids))
	for i, v := range vids {
		vid := strings.TrimSpace(fmt.Sprint(v))
		if vid == "" || vid == "<nil>" {
			continue
		}
		idx := i + 1
		chapters = append(chapters, Chapter{
			ID: ChapterID(seriesID, vid), Source: Source,
			Title: fmt.Sprintf("第%d集", idx), VideoURL: MediaPlaceholder(vid), CurrentEpisode: rawEpisode(idx),
		})
	}
	if len(chapters) == 0 {
		return Detail{}, errors.New("红果详情没有返回剧集 ID")
	}
	drama := DramaFromAny(detail, "短剧")
	if drama.Title == "" {
		drama.Title, drama.Name = title, title
	}
	if drama.SourceID == "" {
		drama.ID, drama.Source, drama.SourceID = DramaID(seriesID), Source, seriesID
	}
	return Detail{Drama: drama, Chapters: chapters}, nil
}

// SocialInfo 是网页详情里的社交数据。
//
// **App 详情接口没有这些字段。** App 的 `/novel/player/video_detail/v1/` 返回 39 个字段，
// 键名含 score/rating/heat/hot 的只有一个 `hot_score`——评分只能从网页拿。
type SocialInfo struct {
	// Rating 是评分，如 9.2；HasRating 为 false 表示这部剧本就没评分
	// （实测 5 部里 2 部没有），不是解析失败。
	Rating    float64
	HasRating bool
	// RatingCount 是评分人数，如 "33371"（上游给的是字符串）。
	RatingCount string
}

// ParseWebSocialInfo 从网页详情 HTML 里取社交数据——目前只有评分。
//
// 注意 `seriesSocialInfo` 与 `seriesDetail` 是**同级**字段，不在它里面；写进
// `seriesDetail` 里查会永远查不到。
func ParseWebSocialInfo(body string) SocialInfo {
	page := RouterLoaderMap(ParseRouterData(body), "detail_page", "detail_")
	social := nestedMap(page, "seriesSocialInfo")
	if len(social) == 0 {
		return SocialInfo{}
	}
	info := SocialInfo{RatingCount: mapString(social, "rating_count")}
	// ParseRouterData 走 UseNumber，所以这里拿到的是 json.Number，不是 float64；
	// 用 fmt.Sprint 一并兜住两种类型。
	if rating, err := strconv.ParseFloat(fmt.Sprint(social["rating"]), 64); err == nil && rating > 0 {
		info.Rating, info.HasRating = rating, true
	}
	return info
}

// FetchWebSocialInfo 取网页详情的社交数据（评分）。
//
// **取不到就当没有**，不返回 error：评分是锦上添花，不该因为它拖垮整个详情页。
func (c *Client) FetchWebSocialInfo(ctx context.Context, seriesID string) SocialInfo {
	if !NumericID.MatchString(seriesID) {
		return SocialInfo{}
	}
	body, err := c.FetchText(ctx, c.webBase()+"/detail?series_id="+url.QueryEscape(seriesID), c.webBase()+"/")
	if err != nil {
		return SocialInfo{}
	}
	return ParseWebSocialInfo(body)
}

func (c *Client) FetchDetail(ctx context.Context, seriesID string) (Detail, error) {
	seriesID = strings.TrimPrefix(strings.TrimSpace(seriesID), "hg-series-v1:")
	if !NumericID.MatchString(seriesID) {
		return Detail{}, errors.New("红果剧集 ID 无效")
	}
	result, appErr := c.AppRequest(ctx, http.MethodPost, "/novel/player/video_detail/v1/", nil, map[string]any{"series_id": seriesID})
	if appErr == nil {
		if entry, err := ParseAppDetail(result, seriesID); err == nil {
			return entry, nil
		} else {
			appErr = err
		}
	}
	if ctx.Err() != nil {
		return Detail{}, ctx.Err()
	}
	body, webErr := c.FetchText(ctx, c.webBase()+"/detail?series_id="+url.QueryEscape(seriesID), c.webBase()+"/")
	if webErr == nil {
		return ParseWebDetail(body, seriesID)
	}
	return Detail{}, fmt.Errorf("App 详情失败: %v；网页详情失败: %w", appErr, webErr)
}

File: packages/hongguo_dart/lib/src/detail.dart
import 'client.dart';
import 'ids.dart';
import 'json.dart';
import 'web.dart';

/// 从 App 详情响应里解析出剧与分集。对照 Go 的 `ParseAppDetail`。
///
/// 校验比 Go 更硬的地方：那边返回部分结果 + error，调用方会丢掉；这里直接抛。
DramaDetail parseAppDetail(Map<String, dynamic> result, String seriesId) {
  final detail = nestedMap(result, const ['data', 'video_data']);
  final returnedId = mapString(detail, const ['series_id_str', 'series_id']);
  if (returnedId != seriesId) {
    throw HongguoRequestException('红果 App 未返回所请求的剧集');
  }

  final drama = dramaFromAny(detail, category: '短剧');
  final seenVideos = <String>{};
  final seenNumbers = <int>{};
  final episodes = <Episode>[];

  for (final row in anyList(detail?['video_list'])) {
    if (row is! Map<String, dynamic>) continue;
    final videoId = mapString(row, const ['vid']);
    final number = int.tryParse(mapString(row, const ['vid_index']));
    if (number == null || number < 1 || !numericIdPattern.hasMatch(videoId)) {
      throw HongguoRequestException('红果 App 分集编号或视频 ID 无效');
    }
    final owner = mapString(row, const ['series_id']);
    if (owner.isNotEmpty && owner != seriesId) {
      throw HongguoRequestException('红果 App 返回了其他剧集的分集');
    }
    if (!seenVideos.add(videoId) || !seenNumbers.add(number)) {
      throw HongguoRequestException('红果 App 返回了重复分集');
    }
    episodes.add(
      Episode(
        id: episodeId(seriesId, videoId),
        number: number,
        title: '第$number集',
        videoId: videoId,
        mediaUrl: mediaPlaceholder(videoId),
      ),
    );
  }

  episodes.sort((left, right) => left.number.compareTo(right.number));

  final total = int.tryParse(mapString(detail, const ['episode_cnt'])) ?? 0;
  if (episodes.isEmpty || total > episodes.length) {
    throw HongguoRequestException('红果 App 未返回完整分集，已尝试其他详情入口');
  }
  for (var index = 0; index < episodes.length; index++) {
    if (episodes[index].number != index + 1) {
      throw HongguoRequestException('红果 App 分集列表不连续');
    }
  }
  return DramaDetail(drama: drama, episodes: episodes);
}

/// 从网页详情 HTML 里解析出剧与分集。对照 Go 的 `ParseWebDetail`。
///
/// 这是 [HongguoDetailApi.fetchDetail] 的**第二条腿**，也是现在唯一还能用的那条：
/// 上游 `/novel/player/video_detail/v1/` 已失效（HTTP 200 但响应体为空），而网页
/// `_ROUTER_DATA` 里的 `seriesDetail` 仍然完整——标题、简介、标签、分类、集数与
/// `vid_list` 都在，且 `vid_list` 的长度与分类接口给的集数逐部一致（实测 6 部）。
///
/// 网页**没有播放量**（也没有热度）。调用方要播放量的话，用 [mergeDrama] 把分类快照
/// 并回来，不要在这里编一个。
///
/// 与 Go 的差别：网页给了 `series_id` 就核对，对不上直接抛。Go 不核对，宁可报
/// 「返回了其他剧集」也别把别的剧的分集挂到这部剧上。
DramaDetail parseWebDetail(String body, String seriesId) {
  final page = routerLoaderMap(
    parseRouterData(body),
    const ['detail_page', 'detail_'],
  );
  final detail = nestedMap(page, const ['seriesDetail']);
  if (detail == null || detail.isEmpty) {
    throw HongguoRequestException('红果网页详情为空，页面结构可能已变化');
  }
  final returnedId = mapString(detail, const ['series_id', 'series_id_str']);
  if (returnedId.isNotEmpty && returnedId != seriesId) {
    throw HongguoRequestException('红果网页详情返回了其他剧集');
  }

  final episodes = <Episode>[];
  final vids = anyList(detail['vid_list']);
  for (var index = 0; index < vids.length; index++) {
    final raw = vids[index];
    final videoId = raw == null ? '' : '$raw'.trim();
    // 网页只给一串 vid、不给集号，所以集号 = **下标 + 1**。任何一项不是合法 vid
    // 就整页判不可用：悄悄跳过会让后面的集号整体前移，症状是「选集里第 7 集其实是
    // 第 8 集」，比直接报错难查得多。
    if (!numericIdPattern.hasMatch(videoId)) {
      throw HongguoRequestException('红果网页详情含无效视频 ID');
    }
    final number = index + 1;
    episodes.add(
      Episode(
        id: episodeId(seriesId, videoId),
        number: number,
        title: '第$number集',
        videoId: videoId,
        mediaUrl: mediaPlaceholder(videoId),
      ),
    );
  }
  if (episodes.isEmpty) {
    throw HongguoRequestException('红果网页详情没有返回剧集 ID');
  }

  var drama = dramaFromAny(detail, category: '短剧');
  if (drama.title.isEmpty || drama.sourceId.isEmpty) {
    drama = mergeDrama(
      drama,
      Drama(
        id: dramaId(seriesId),
        source: hongguoSource,
        sourceId: seriesId,
        title: firstNonEmpty([
          mapString(detail, const ['series_name', 'series_title', 'name']),
          seriesId,
        ]),
      ),
    );
  }
  return DramaDetail(drama: drama, episodes: episodes);
}

/// 网页详情里的社交数据。对照 Go 的 `SocialInfo`。
///
/// **App 详情接口没有这些字段。** App 的 `/novel/player/video_detail/v1/` 返回 39 个字段，
/// 键名含 `score`/`rating`/`heat`/`hot` 的只有一个 `hot_score`——评分只能从网页拿。
class SocialInfo {
  const SocialInfo({this.rating = 0, this.ratingCount = ''});

  /// 评分，如 9.2。**0 表示这部剧本就没评分**（实测 5 部里 2 部没有），不是解析失败。
  final double rating;

  /// 评分人数，如 `33371`（上游给的是字符串）。
  final String ratingCount;

  bool get hasRating => rating > 0;
}

/// 从网页详情 HTML 里取社交数据——目前只有评分。
///
/// 注意 `seriesSocialInfo` 与 `seriesDetail` 是**同级**字段，不在它里面；写进
/// `seriesDetail` 里查会永远查不到。
///
/// 对照 Go 的 `ParseWebSocialInfo`。
SocialInfo parseWebSocialInfo(String body) {
  final page = routerLoaderMap(
    parseRouterData(body),
    const ['detail_page', 'detail_'],
  );
  final social = nestedMap(page, const ['seriesSocialInfo']);
  if (social == null) return const SocialInfo();

  final rating = (social['rating'] as num?)?.toDouble() ?? 0;
  return SocialInfo(
    rating: rating > 0 ? rating : 0,
    ratingCount: mapString(social, const ['rating_count']),
  );
}

/// 详情相关的客户端方法。对照 Go `hongguo/detail.go` 的 `Client.FetchDetail`。
extension HongguoDetailApi on HongguoClient {
  /// 网页详情页的 URL。**详情与评分是同一个页面**——`seriesDetail` 与
  /// `seriesSocialInfo` 是同级字段，不在它里面。
  ///
  /// 所以取详情与取评分打的是同一个地址；`/detail` 在 TTL 缓存白名单里
  /// （`lib/data/network.dart`），第二次会直接命中，不会重复下这约 290 KB。
  ///
  /// 用 `webBase` 而不是常量——那边是可被覆盖的站点根（同 [fetchWebCategoryPage]）。
  /// 之前这里读的是常量，于是覆盖了 `webBase` 的调用方会发现详情走了覆盖后的主机、
  /// 评分却还打在默认主机上。
  String webDetailUrl(String seriesId) =>
      '${trimTrailingSlash(webBase)}/detail'
      '?series_id=${Uri.encodeQueryComponent(seriesId)}';

  /// 网页详情页要带的 Referer（站点根）。
  String get webDetailReferer => '${trimTrailingSlash(webBase)}/';

  /// 取详情 + 分集。
  ///
  /// **两条腿：App 签名接口 → 网页详情。** 顺序与 Go 的 `FetchDetail` 一致。
  ///
  /// 为什么必须有第二条：上游 `/novel/player/video_detail/v1/` 已经失效——HTTP 200
  /// 但响应体是空的（实测换 host、GET/POST、加 `book_id` 全是空体；`/video_detail/`
  /// 则返回 404，说明路径还在、只是不再吐数据）。只留 App 那条腿的话，点开任何一部剧
  /// 都会报「红果 App 接口未返回有效数据」。网页那条腿的数据是完整的（见
  /// [parseWebDetail]），所以它不是「降级方案」，是眼下唯一可用的源；上游哪天把 App
  /// 详情放回来，这里会自动用回它。
  ///
  /// 网页里**没有播放量**。要显示播放量的调用方，用 [mergeDrama] 把分类快照并回来。
  Future<DramaDetail> fetchDetail(String seriesId) async {
    final normalized =
        seriesId.trim().replaceFirst(RegExp(r'^hg-series-v1:'), '');
    if (!numericIdPattern.hasMatch(normalized)) {
      throw HongguoRequestException('红果剧集 ID 无效');
    }

    Object? appError;
    try {
      final result = await appRequest(
        method: 'POST',
        path: '/novel/player/video_detail/v1/',
        payload: {'series_id': normalized},
      );
      return parseAppDetail(result, normalized);
    } catch (error) {
      // 空体、格式异常、业务码非 0、以及 HTTP 4xx 都落这里——每一种都该试网页。
      appError = error;
    }

    try {
      final body = await fetchText(
        webDetailUrl(normalized),
        referer: webDetailReferer,
      );
      return parseWebDetail(body, normalized);
    } catch (webError) {
      // 两条腿的原因都带上——只报后一条的话，前一条为什么失败就永远查不到了。
      throw HongguoRequestException(
        '红果 App 详情失败：$appError；网页详情失败：$webError',
      );
    }
  }

  /// 取网页详情的社交数据（评分）。对照 Go 的 `Client.FetchWebSocialInfo`。
  ///
  /// **取不到就当没有**，不抛：评分是锦上添花，不该因为它拖垮整个详情页。网页是
  /// 免签名的，页大小约 290 KB，但 `/detail` 在 TTL 缓存白名单里（见
  /// `lib/data/network.dart`），而 [fetchDetail] 打的是**同一个地址**——所以
  /// 详情那条腿落到网页之后，这次调用基本都是一次缓存命中。
  Future<SocialInfo> fetchWebSocialInfo(String seriesId) async {
    final normalized =
        seriesId.trim().replaceFirst(RegExp(r'^hg-series-v1:'), '');
    if (!numericIdPattern.hasMatch(normalized)) return const SocialInfo();
    try {
      final body = await fetchText(
        webDetailUrl(normalized),
        referer: webDetailReferer,
      );
      return parseWebSocialInfo(body);
    } on HongguoRequestException {
      // fetchText 的传输 / 状态码 / 拦截页失败都抛这个。评分缺席不该让详情页出错。
      return const SocialInfo();
    }
  }
}
